Upsert fails with parameters out of order

After writing an upsert mutation, I was surprised to see a 500 from the Prisma server. Looking in the Docker logs, it seems like the parameters for the SQL query are not relayed in order. Specifically, the `type` variable is being assigned to the `id` field in the database, which triggers a uniqueness constraint error.

It appears that `type` is getting pushed to the end of the SELECT clause parameters, even though it's listed as the 4th parameter in the INSERT clause. Could it be related to type being a custom enum, or maybe even its position in the original schema definition?

Type definition:

```
type Permission {
  id: ID! @unique
  name: String! @unique
  crudKey: String! @unique
  type: PermissionType!

  resourceName: String
  allowedOperation: CrudOperation

  customOperation: CustomOperation @relation(name: "CustomOperationOnPermission")

  users: [User!]! @relation(name: "UserOnPermissions")
  groups: [PermissionGroup!]! @relation(name: "PermissionGroupOnPermissions")
}
```

Query:

```js
const permission = {
  resourceName: 'Node',
  allowedOperation: 'UPDATE',
  key: 'Node:UPDATE'
};

context.db.mutation.upsertPermission({ 
  create: {
    name: permission.key,
    crudKey: permission.key,
    type: 'CRUD',
    resourceName: permission.resourceName,
    allowedOperation: permission.allowedOperation,
  },
  update: {
    resourceName: permission.resourceName,
    allowedOperation: permission.allowedOperation,
  },    
  where: {     
    crudKey: permission.key,
  },
})
```

Error in Docker logs:

```
[Bugsnag - local / testing] Error report: com.bugsnag.Report@32174158
Caused by: java.sql.SQLException: Duplicate entry 'CRUD' for key 'PRIMARY'
Query is: INSERT INTO `prisma-admin@dev`.`Permission` ( `name` , `allowedOperation` , `crudKey` , `type` , `resourceName` , `id` ) SELECT  ? , ? , ? , ? , ? , ? FROM DUAL where not exists (
select `id` from `prisma-admin@dev`.`Permission` where `crudKey` = ? limit 1);, parameters ['Node:UPDATE','UPDATE','Node:UPDATE','Node','cjcznwwpu00ce01675iv1ijqf','CRUD','Node:UPDATE']
        at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:146)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:217)
        at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:218)
        ... 24 more
java.sql.SQLIntegrityConstraintViolationException: (conn=37) Duplicate entry 'CRUD' for key 'PRIMARY'
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:157)
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:106)
        at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:235)
        at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:224)
        at org.mariadb.jdbc.MariaDbPreparedStatementClient.execute(MariaDbPreparedStatementClient.java:159)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
        at slick.jdbc.StatementInvoker.results(StatementInvoker.scala:39)
        at slick.jdbc.StatementInvoker.iteratorTo(StatementInvoker.scala:22)
        at slick.jdbc.Invoker.first(Invoker.scala:31)
        at slick.jdbc.Invoker.first$(Invoker.scala:30)
[Bugsnag - local / testing] Error report: com.bugsnag.Report@66ca7960
        at slick.jdbc.StatementInvoker.first(StatementInvoker.scala:16)
        at slick.jdbc.StreamingInvokerAction$HeadAction.run(StreamingInvokerAction.scala:52)
        at slick.jdbc.StreamingInvokerAction$HeadAction.run(StreamingInvokerAction.scala:51)
        at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:240)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.collection.IterableLike.foreach(IterableLike.scala:71)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:240)
        at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:238)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:240)
        at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:240)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[Bugsnag - local / testing] Error report: com.bugsnag.Report@3aa0309c
Caused by: java.sql.SQLException: Duplicate entry 'CRUD' for key 'PRIMARY'
Query is: INSERT INTO `prisma-admin@dev`.`Permission` ( `name` , `allowedOperation` , `crudKey` , `type` , `resourceName` , `id` ) SELECT  ? , ? , ? , ? , ? , ? FROM DUAL where not exists (
select `id` from `prisma-admin@dev`.`Permission` where `crudKey` = ? limit 1);, parameters ['Node:DELETE','DELETE','Node:DELETE','Node','cjcznwwro00cn0167tc089duu','CRUD','Node:DELETE']
        at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:146)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:217)
        at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:218)
        ... 24 more
```

--- a/server/api/src/main/scala/com/prisma/api/database/DatabaseMutationBuilder.scala
+++ b/server/api/src/main/scala/com/prisma/api/database/DatabaseMutationBuilder.scala
@@ -31,11 +31,13 @@ object DatabaseMutationBuilder {
 
   def createDataItemIfUniqueDoesNotExist(projectId: String,
                                          where: NodeSelector,
-                                         createArgs: CoolArgs): SqlStreamingAction[Vector[Int], Int, Effect]#ResultAction[Int, NoStream, Effect] = {
-    val escapedColumns = combineByComma(createArgs.raw.keys.map(escapeKey))
-    val insertValues   = combineByComma(createArgs.raw.values.map(escapeUnsafeParam))
-    (sql"INSERT INTO `#${projectId}`.`#${where.model.name}` (" ++ escapedColumns ++ sql")" ++
-      sql"SELECT " ++ insertValues ++
+                                         args: CoolArgs): SqlStreamingAction[Vector[Int], Int, Effect]#ResultAction[Int, NoStream, Effect] = {
+
+    val escapedKeyValueTuples = args.raw.toList.map(x => (escapeKey(x._1), escapeUnsafeParam(x._2)))
+    val escapedKeys           = combineByComma(escapedKeyValueTuples.map(_._1))
+    val escapedValues         = combineByComma(escapedKeyValueTuples.map(_._2))
+    (sql"INSERT INTO `#${projectId}`.`#${where.model.name}` (" ++ escapedKeys ++ sql")" ++
+      sql"SELECT " ++ escapedValues ++
       sql"FROM DUAL" ++
       sql"where not exists (select `id` from `#${projectId}`.`#${where.model.name}` where `#${where.field.name}` = ${where.fieldValue} limit 1);").asUpdate
   }

